### PR TITLE
Add support for Dictionary to FieldLoader and FieldSaver

### PR DESCRIPTION
**Pull request by penev92**
Inspired by #9181, this makes `FieldLoader` capable of loading `Dictionary<,>` directly, instead of traits relying on their own code to load them in a number of different ways.
Also `FieldSaver` is not left out, and is able to save them as YAML (for saving maps) and potentially as string, for trait documentation generation.

As usual, split into nice small easy to review commits.

Easiest way to test this is with my `PlugInit` PR, because that uses both the `FieldLoader` and `FieldSaver` changes.

### CODE REVIEW

File: OpenRA.Game/FieldLoader.cs [hunk 1]
```diff
@@ -441,6 +442,20 @@ public static object GetValue(string fieldName, Type fieldType, string value, Me
 					addMethod.Invoke(set, new[] { GetValue(fieldName, fieldType.GetGenericArguments()[0], parts[i].Trim(), field) });
 				return set;
 			}
+			else if (fieldType.IsGenericType && fieldType.GetGenericTypeDefinition() == typeof(Dictionary<,>))
+			{
+				var dict = Activator.CreateInstance(fieldType);
+				var addMethod = fieldType.GetMethod("Add", fieldType.GetGenericArguments());
```
Comment (pchote) [inline OpenRA.Game/FieldLoader.cs]: You call `fieldType.GetGenericArguments()` `2 * <number of children> + 1` times.

File: OpenRA.Mods.Common/Traits/GainsExperience.cs [hunk 2]
```diff
@@ -20,7 +19,6 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("This actor's experience increases when it has killed a GivesExperience actor.")]
 	public class GainsExperienceInfo : ITraitInfo, Requires<ValuedInfo>, Requires<UpgradeManagerInfo>
 	{
-		[FieldLoader.LoadUsing("LoadUpgrades", true)]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/GainsExperience.cs]: The `true` indicates that this is a required field.  You will need to add the appropriate attribute to not change behavior here.

File: OpenRA.Mods.Common/Traits/Infantry/TerrainModifiesDamage.cs [hunk 3]
```diff
@@ -17,7 +16,6 @@ namespace OpenRA.Mods.Common.Traits
 {
 	public class TerrainModifiesDamageInfo : ITraitInfo
 	{
-		[FieldLoader.LoadUsing("LoadPercents", true)]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Infantry/TerrainModifiesDamage.cs]: And here.

File: OpenRA.Game/FieldLoader.cs (hunk 1 above)
Comment (penev92) [inline OpenRA.Game/FieldLoader.cs]: OK, does that give me anything, because the current setup is a lot clearer than that would be I think.

**Comment by penev92** [comment]
Added `[FieldLoader.Require]` to not change behaviour and avoided calling `fieldType.GetGenericArguments()` repeatedly.

**Comment by Mailaender** [comment]
Nice cleanup and can't spot regressions. :white_check_mark:

File: OpenRA.Game/FieldLoader.cs [hunk 4]
```diff
@@ -588,7 +604,7 @@ static FieldLoadInfo[] BuildTypeLoadInfo(Type type)
 
 				var loader = sa.GetLoader(type);
 				if (loader == null && sa.FromYamlKey)
-					loader = (yaml) => GetValue(yamlName, field.FieldType, yaml.Value, field);
+					loader = (yaml) => GetValue(yamlName, field.FieldType, yaml, field);
```
Comment (RoosterDragon) [inline OpenRA.Game/FieldLoader.cs]: Just since we're here:
```(yaml)``` -> ```yaml```.

**Comment by RoosterDragon** [comment]
:+1:

**Pull request merged**